scripts/convert_data.py

```diff
@@ -0,0 +1,243 @@
+#!/usr/bin/env python
+
+# Data conversion utility for ner-uk data set. Originally created for purposes of training Stanza model.
+# Later Stanza has merged some of this code inside.
+# Terminology:
+# IOB https://en.wikipedia.org/wiki/Inside%E2%80%93outside%E2%80%93beginning_(tagging) Some sources refer to this as BIO
+# BEIOS - extended IOB with E == end (for multi word tokens)
+
+
+import argparse
+import logging
+import os
+import glob
+import time
+from collections import namedtuple
+import re
+from typing import Tuple
+
+from tqdm import tqdm
+from random import choices, shuffle
+
+BsfInfo = namedtuple('BsfInfo', 'id, tag, start_idx, end_idx, token')
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.INFO)
+
+
+def format_token_as_beios(token: str, tag: str) -> list:
+    t_words = token.split()
+    res = []
+    if len(t_words) == 1:
+        res.append(token + ' S-' + tag)
+    else:
+        res.append(t_words[0] + ' B-' + tag)
+        for t_word in t_words[1: -1]:
+            res.append(t_word + ' I-' + tag)
+        res.append(t_words[-1] + ' E-' + tag)
+    return res
+
+
+def format_token_as_iob(token: str, tag: str) -> list:
+    t_words = token.split()
+    res = []
+    if len(t_words) == 1:
+        res.append(token + ' B-' + tag)
+    else:
+        res.append(t_words[0] + ' B-' + tag)
+        for t_word in t_words[1:]:
+            res.append(t_word + ' I-' + tag)
+    return res
+
+
+def convert_bsf(data: str, bsf_markup: str, converter: str = 'beios') -> str:
+    """
+    Convert data file with NER markup in Brat standoff Format to BEIOS or IOB format.
+
+    :param converter: iob or beios converter to use for document
+    :param data: tokenized data to be converted. Each token separated with a space
+    :param bsf_markup: Brat standoff Format markup
+    :return: data in BEIOS or IOB format https://en.wikipedia.org/wiki/Inside–outside–beginning_(tagging)
+    """
+
+    def join_simple_chunk(chunk: str) -> list:
+        if len(chunk.strip()) == 0:
+            return []
+        # keep the newlines, but discard the non-newline whitespace
+        tokens = re.split(r'(\n)|\s', chunk.strip())
+        # the re will return None for splits which were not caught in a group
+        tokens = [x for x in tokens if x is not None]
+        return [token + ' O' if len(token.strip()) > 0 else token for token in tokens]
+
+    converters = {'beios': format_token_as_beios, 'iob': format_token_as_iob}
+    res = []
+    markup = parse_bsf(bsf_markup)
+
+    prev_idx = 0
+    m_ln: BsfInfo
+    for m_ln in markup:
+        res += join_simple_chunk(data[prev_idx:m_ln.start_idx])
+
+        convert_f = converters[converter]
+        res.extend(convert_f(m_ln.token, m_ln.tag))
+        prev_idx = m_ln.end_idx
+
+    if prev_idx < len(data) - 1:
+        res += join_simple_chunk(data[prev_idx:])
+
+    return '\n'.join(res)
+
+
+def parse_bsf(bsf_data: str) -> list:
```

Comment:
Any good ideas how to keep this function and `BsfInfo` definition in one place? Separate package looks like an overkill, of course.

Reply:
What makes me a bit nervous is that complex regex, that we now have in two different packages. On the other hand I don't think it's going to change anytime soon :)

Reply:
Maybe (just maybe), the whole logic to convert between formats can be separated as a package, which has some value per se (i.e might be used by other people around the world).

Reply:
I also spent quite a bit of time debating myself whether I should package this as a library. Not a fan of duplicating code but ended up with the following:
- so far this code is generally unique to our 2 repos
- adding an external dependency because 1 function might be an overkill and complicate dependency management and making changes in that shared code
- makes a bit harder to read and understand this relatively simple code
Happy to package it as soon as we have a need for 3rd usage :)

Reply:
Ok, works for me